File: Data Structures & Algorithms/non-overlapping-intervals/submission-9.py
class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        # return the min numbers of intervals we need to remove to make the rest
        # of the intervals non-overlapping

        # we want the smalles removal

        intervals.sort(key=lambda x:x[0])

        stk = []

        for start, end in intervals:
            # empty
            if not stk: 
                stk.append([start, end])

            # end is smalller than prev one 
            elif end < stk[-1][1]:
                stk.pop()
                stk.append([start, end])
            
            # end is bigger but we do not know if the start: overlaps with list in stk
            elif start == stk[-1][1]:
                stk.append([start, end])
            
            elif start < stk[-1][1]:
                continue
            
            else:
                stk.append([start, end])


        return len(intervals) - len(stk)

File: Data Structures & Algorithms/validate-parentheses/submission-9.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> mapping = {
            {')', '('},
            {']', '['},
            {'}', '{'}
        };

        for (char c : s) {
            if (mapping.count(c)) {
                if (!stk.empty() && mapping[c] == stk.top()) {
                    stk.pop();
                } else {
                    return false;
                }
            } else {
                stk.push(c);
            }
        }

        return stk.empty();
    }
};
